core/common/buffer_view.hpp

```diff
@@ -0,0 +1,160 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef KAGOME_BUFFERVIEW_HPP
+#define KAGOME_BUFFERVIEW_HPP
+
+#include <gsl/span>
+
+#include "common/hexutil.hpp"
+
+namespace kagome::common {
+  template <size_t MaxSize>
+  class BufferT;
+}
+
+namespace kagome::common {
+
+  class BufferView : public gsl::span<const uint8_t>
+  {
+    using Span = gsl::span<const uint8_t>;
+    std::optional<std::vector<Span::value_type>> holder_{};
```

Comment:
`Buffer` is `vector`.  
`BufferView` is `span`.  
`BufferOrView` is `variant<vector, span>` (with convenience methods/casts).  

Like [rust's `Cow`](https://doc.rust-lang.org/std/borrow/enum.Cow.html).

Example from [fuhon](https://github.com/filecoin-project/cpp-filecoin/blob/16ffecde801891f36b4a4168bfb3021510cc03d4/core/common/bytes_cow.hpp#L13-L15).

```suggestion
```